ytt/001-schemas/README.md

```diff
@@ -0,0 +1,1647 @@
+
+Keep in mind that we aim for being able to articulate schema for any YAML document `ytt` handles. e.g. [ytt#103@schema-for-templates](https://github.com/k14s/ytt/issues/103#issuecomment-624326922)
+
+## Sources of Inspiration
+
+Be aware of prior art: avoid reinvention; shamelessly "repurpose" great ideas.
+
+- [JSON Schema](https://json-schema.org/)
+- [OpenAPI v3 schema](https://swagger.io/specification/) — _(see also [ytt#103@openapi](https://github.com/k14s/ytt/issues/103#issuecomment-672296576))_
+
+Remember the differences in context:
+
+- we seek simplicity, brevity, and readability.
+
+
+# Specification
+## Glossary
+
+`ytt` bridges two type spaces: YAML and Starlark. Terms from both spaces are used in this specification
+
+When referring to elements that are YAML, this specification uses terms defined in the [YAML 1.2 specification](https://yaml.org/spec/1.2/spec.html).
+
+Commonly used terms:
+- **document** — root node with a single value that can be a mapping, sequence, or scalar.
+- **mapping** — a map (aka "dictionary") containing zero or more key/value pairs. Each value can be a mapping, sequence, or scalar.
+- **node** - the composite of a YAML document; one of: a document, map, map item, array, array item, or scalar.
+- **scalar** — single value type: boolean, float, integer, null, string.
+- **sequence** — an array (aka "list") containing zero or more elements, in order.
+
+
+Terms as they are meant in this specification:
+
+- **schema** — (n.) a definition of the structure, types and validation rules for a given YAML document.
+- **type** — (n.) a specific shape for a given node: a mapping, sequence, or scalar.
+- **type checking** — (v.) comparing the shape of a YAML structure against its schema's types.
+- **type check** — (n.) the result of type checking
+- **validating** — (v.) determining whether the values contained in a YAML structure are acceptable.
+- **validation** — (n.) the result of validating
+- **validation rule** — predicate that indicates whether a given (sub-)structure is valid or not.
+- **verifying** — (v.) a short-hand way of saying, "type checking and validation."
+- **violation** — (n.) an instance of either a failed type check or a failed validation.
+
+## Part 1: Declaring Schema
+
+Configuration Authors establish a Schema by capturing the structure in a YAML document:
+
+### @schema/definition
+
+```yaml
+#@schema/definition name="data/values", version="1.0.0", apply_by=pass
+---
+#! Schema contents
+```
+
+- `name=` sets an identifier for this schema that can be referenced later.
+  - the name `data/values` is reserved; it identifies the Data Values.
+  - if no name kwarg is given, "`data/values`" is assumed.
+- `version=` sets the version number of the schema, providing a means of orderly upgrades
+  (we strongly recommend using [Semantic Versioning 2.0](https://semver.org/) semantics). 
```

Comment:
Remove: `version` is unnecessary. Products versions ought to be sensitive to changes in schema. If, for example, CF for K8s makes a change in their schema, that should result in a product version bump. If a Configuration Author packages their product up as a `imgpkg` bundle, _that bundle_ is versioned; and if there are schema changes, _that bundle_'s version should be bumped.